File: assetcore/tests/_helpers/paths.py
"""SSoT đường dẫn cho MỌI test BE đọc đĩa.

Vì sao có file này (đọc trước khi sửa)
--------------------------------------
Class-of-bug đóng ở đây là **guard xanh giả**. Một guard quét
``assetcore/assetcore/workflow/*.json`` bằng đường dẫn tính theo **ĐỘ SÂU**
(``Path(__file__).resolve().parents[1]``). Khi file test bị dời xuống một thư
mục con — đúng việc lô B2/B3 sắp làm — ``parents[1]`` trỏ sai một cấp, bộ quét
trả về **0 file**, và mọi khẳng định dạng "không có vi phạm nào" trở thành đúng
một cách **rỗng tuếch** trong khi suite vẫn XANH.

Backend nguy hiểm hơn frontend ở chỗ này: FE có ``vue-tsc`` bắt 100% lỗi dời
import, **Python không có compiler nào bắt được**. Lưới an toàn duy nhất là
chạy suite thật — mà suite BE hiện **không tất định** (đo 2026-08-13: cùng cây
mã cho ``errors=27`` rồi ``errors=24``). Nên phòng thủ phải nằm ngay trong lớp
đường dẫn.

Hai lớp phòng thủ
-----------------
1. Mọi hằng số neo bằng ``frappe.get_app_path`` hoặc mốc thư mục — **không đếm
   số cấp ``..``**. File này dời đi đâu thì đường dẫn vẫn đúng.
2. :func:`require_dir` **ném lỗi ngay lúc import** nếu thư mục neo biến mất, và
   :func:`list_files` ném lỗi nếu quét ra ít file hơn ngưỡng. Thư mục dời đi ⇒
   suite ĐỎ ầm ĩ, không bao giờ "đếm 0 rồi PASS".

Quy tắc bắt buộc (SPEC BE §5.2 N5)
----------------------------------
* Mọi test đọc đĩa **phải** lấy đường dẫn từ module này.
  **CẤM** ``Path(__file__).resolve().parents[N]`` và
  ``os.path.dirname(os.path.dirname(__file__))``.
* Mọi guard quét thư mục **phải** chốt dân số tối thiểu — dùng
  ``list_files(DIR, ".json", min_count=N)`` hoặc
  ``self.assertGreater(len(files), N)``.
"""

from __future__ import annotations

import os
from typing import Iterable

import frappe

APP = "assetcore"


def require_dir(path: str, label: str) -> str:
    """Trả về ``path`` nếu là thư mục có thật; ném lỗi ầm ĩ nếu không.

    Đây là chốt chặn xanh giả: thư mục neo biến mất phải làm suite ĐỎ ngay lúc
    import, chứ không để bộ quét trả mảng rỗng rồi mọi assertion "không có vi
    phạm" thành đúng-rỗng-tuếch.
    """
    if not os.path.isdir(path):
        raise RuntimeError(
            f"[tests/_helpers/paths.py] Thư mục neo {label} không tồn tại: {path}\n"
            "Thư mục đã bị dời/đổi tên. Sửa paths.py + test liên quan. "
            "KHÔNG được để test quét vào hư vô rồi báo PASS."
        )
    return path


def require_file(path: str, label: str) -> str:
    """Như :func:`require_dir` nhưng cho file."""
    if not os.path.isfile(path):
        raise RuntimeError(
            f"[tests/_helpers/paths.py] File neo {label} không tồn tại: {path}"
        )
    return path


#: ``apps/assetcore/assetcore`` — gốc package Python của app.
APP_ROOT = require_dir(frappe.get_app_path(APP), "APP_ROOT")
#: ``apps/assetcore`` — gốc repo (cha của package, chứa cả ``frontend/`` và ``docs/``).
REPO_ROOT = require_dir(os.path.dirname(APP_ROOT), "REPO_ROOT")

TESTS_DIR = require_dir(os.path.join(APP_ROOT, "tests"), "TESTS_DIR")
SERVICES_DIR = require_dir(os.path.join(APP_ROOT, "services"), "SERVICES_DIR")
SHARED_DIR = require_dir(os.path.join(SERVICES_DIR, "shared"), "SHARED_DIR")
API_DIR = require_dir(os.path.join(APP_ROOT, "api"), "API_DIR")
UTILS_DIR = require_dir(os.path.join(APP_ROOT, "utils"), "UTILS_DIR")
SETUP_DIR = require_dir(os.path.join(APP_ROOT, "setup"), "SETUP_DIR")
PATCHES_DIR = require_dir(os.path.join(APP_ROOT, "patches"), "PATCHES_DIR")
SCRIPTS_DIR = require_dir(os.path.join(APP_ROOT, "scripts"), "SCRIPTS_DIR")
FIXTURES_DIR = require_dir(os.path.join(APP_ROOT, "fixtures"), "FIXTURES_DIR")

#: Module Frappe: ``assetcore/assetcore`` — chứa doctype/ workflow/ workspace/.
MODULE_ROOT = require_dir(os.path.join(APP_ROOT, APP), "MODULE_ROOT")
DOCTYPE_DIR = require_dir(os.path.join(MODULE_ROOT, "doctype"), "DOCTYPE_DIR")
WORKFLOW_DIR = require_dir(os.path.join(MODULE_ROOT, "workflow"), "WORKFLOW_DIR")

#: Ngoài package Python — cùng cấp repo.
DOCS_DIR = require_dir(os.path.join(REPO_ROOT, "docs"), "DOCS_DIR")
FRONTEND_DIR = require_dir(os.path.join(REPO_ROOT, "frontend"), "FRONTEND_DIR")
FRONTEND_SRC = require_dir(os.path.join(FRONTEND_DIR, "src"), "FRONTEND_SRC")


def list_files(
    directory: str,
    suffix: str | tuple[str, ...] | None = None,
    *,
    min_count: int,
    recursive: bool = True,
    skip: Iterable[str] = (),
) -> list[str]:
    """Quét thư mục và **chốt dân số tối thiểu**.

    Ném lỗi nếu ``directory`` không tồn tại hoặc số file quét được nhỏ hơn
    ``min_count``, thay vì trả danh sách rỗng — danh sách rỗng chính là cách
    mọi khẳng định "không tìm thấy vi phạm" trở thành đúng-một-cách-rỗng-tuếch.

    :param min_count: bắt buộc truyền. Đặt bằng số đo từ đĩa, hạ xuống mốc tròn
        để thêm/bớt vài file không gây đỏ giả — nhưng đủ chặt để bắt "thư mục
        biến mất".
    """
    require_dir(directory, os.path.basename(directory))
    skip_set = set(skip)
    out: list[str] = []

    if recursive:
        for root, dirs, files in os.walk(directory):
            dirs[:] = [d for d in dirs if d not in skip_set and d != "__pycache__"]
            for name in files:
                if suffix is None or name.endswith(suffix):
                    out.append(os.path.join(root, name))
    else:
        for name in sorted(os.listdir(directory)):
            full = os.path.join(directory, name)
            if os.path.isfile(full) and (suffix is None or name.endswith(suffix)):
                out.append(full)

    out.sort()
    if len(out) < min_count:
        raise RuntimeError(
            f"[tests/_helpers/paths.py] list_files({os.path.relpath(directory, REPO_ROOT)}) "
            f"chỉ ra {len(out)} file, dưới ngưỡng tối thiểu {min_count}.\n"
            "Thư mục đã bị dời/đổi tên/rỗng đi ⇒ guard đã NGỪNG CANH. "
            "Sửa đường dẫn hoặc cập nhật ngưỡng CÓ CHỦ Ý — đừng để guard đếm 0 rồi PASS."
        )
    return out


def rel_repo(path: str) -> str:
    """Đường dẫn tương đối so với gốc repo — cho thông điệp lỗi đọc được."""
    return os.path.relpath(path, REPO_ROOT)

File: frontend/src/test/paths.ts
// Copyright (c) 2026, AssetCore Team
// SSoT đường dẫn cho MỌI guard đọc đĩa của FE.
//
// ─── Vì sao có file này (đọc trước khi sửa) ───────────────────────────────────
// Class-of-bug đóng ở đây là **guard xanh giả**: một guard quét `src/views` bằng
// đường dẫn tính theo ĐỘ SÂU (`resolve(HERE, '../..')`). Khi file guard bị dời
// sâu thêm một cấp — hoặc khi thư mục bị quét đổi tên — đường dẫn trỏ vào chỗ
// không tồn tại, bộ quét đệ quy trả về **0 file**, mọi khẳng định dạng "không có
// vi phạm nào" thành đúng một cách rỗng tuếch, và suite vẫn **XANH** trong khi
// guard đã ngừng canh. Không có lỗi biên dịch nào bắt được chuyện này:
// `vue-tsc` chỉ phủ đồ thị import, còn lớp guard đọc VĂN BẢN mã nguồn thì nằm
// ngoài tầm compiler.
//
// Hai lớp phòng thủ ở file này:
//   1. `FRONTEND_ROOT` neo bằng cách **đi ngược lên tìm mốc** (`package.json` +
//      `vite.config.ts`), KHÔNG đếm số cấp `..`. File này có bị dời đi đâu thì
//      đường dẫn vẫn đúng.
//   2. `requireDir()` **ném lỗi ngay lúc import** nếu thư mục neo biến mất, và
//      `listFiles()` ném lỗi nếu quét ra ít file hơn ngưỡng. Thư mục dời đi ⇒
//      suite ĐỎ ầm ĩ, không bao giờ "đếm 0 rồi PASS".
//
// ─── Quy tắc bắt buộc (SPEC §5.2 N5/N6) ──────────────────────────────────────
//   • Mọi guard đọc đĩa PHẢI lấy đường dẫn từ file này. CẤM `resolve(HERE, '../..')`.
//   • Mọi guard quét thư mục PHẢI chốt dân số tối thiểu — dùng `listFiles(..., { min })`
//     hoặc `expect(files.length).toBeGreaterThanOrEqual(N)`.
import { existsSync, readdirSync, statSync } from 'node:fs'
import { dirname, join, relative, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

/** Đi ngược lên tìm gốc `frontend/` bằng MỐC, không bằng số cấp `..`. */
function findFrontendRoot(): string {
  let dir = dirname(fileURLToPath(import.meta.url))
  for (let i = 0; i < 12; i += 1) {
    if (existsSync(join(dir, 'package.json')) && existsSync(join(dir, 'vite.config.ts'))) {
      return dir
    }
    const parent = dirname(dir)
    if (parent === dir) break
    dir = parent
  }
  throw new Error(
    '[test/paths.ts] Không tìm thấy gốc `frontend/` (mốc: package.json + vite.config.ts). ' +
      'Cây thư mục đã đổi — sửa file này, ĐỪNG quay lại đường dẫn tính theo độ sâu.',
  )
}

/** Trả về `p` nếu là thư mục có thật; ném lỗi ầm ĩ nếu không — chống xanh giả. */
function requireDir(p: string, label: string): string {
  if (!existsSync(p) || !statSync(p).isDirectory()) {
    throw new Error(
      `[test/paths.ts] Thư mục neo ${label} không tồn tại: ${p}\n` +
        'Thư mục đã bị dời/đổi tên. Sửa paths.ts + guard liên quan. ' +
        'KHÔNG được để guard quét vào hư vô rồi báo PASS.',
    )
  }
  return p
}

export const FRONTEND_ROOT = findFrontendRoot()
/** Gốc app Frappe (`apps/assetcore/`) — cha của `frontend/`. */
export const REPO_ROOT = dirname(FRONTEND_ROOT)

export const SRC = requireDir(join(FRONTEND_ROOT, 'src'), 'SRC')
export const DOCS = requireDir(join(REPO_ROOT, 'docs'), 'DOCS')

export const API = requireDir(join(SRC, 'api'), 'API')
export const COMPONENTS = requireDir(join(SRC, 'components'), 'COMPONENTS')
export const COMPOSABLES = requireDir(join(SRC, 'composables'), 'COMPOSABLES')
export const CONSTANTS = requireDir(join(SRC, 'constants'), 'CONSTANTS')
export const LOCALES = requireDir(join(SRC, 'locales'), 'LOCALES')
export const ROUTER = requireDir(join(SRC, 'router'), 'ROUTER')
export const STORES = requireDir(join(SRC, 'stores'), 'STORES')
export const TYPES = requireDir(join(SRC, 'types'), 'TYPES')
export const UTILS = requireDir(join(SRC, 'utils'), 'UTILS')
export const VIEWS = requireDir(join(SRC, 'views'), 'VIEWS')

/** Nhà của guard đọc đĩa (dựng ở lô L2). */
export const GUARDS = requireDir(join(SRC, 'guards'), 'GUARDS')
/** Nhà của test tích hợp / khởi động / luồng chéo (dựng ở lô L3). */
export const INTEGRATION = requireDir(join(SRC, 'integration'), 'INTEGRATION')

/** Đường dẫn tương đối so với `src/` — dùng cho thông điệp lỗi đọc được. */
export function rel(absPath: string): string {
  return relative(SRC, absPath).split('\\').join('/')
}

/** Đường dẫn tương đối so với gốc app — dùng khi thông điệp vượt ra ngoài `frontend/`. */
export function relRepo(absPath: string): string {
  return relative(REPO_ROOT, absPath).split('\\').join('/')
}

export interface ListFilesOptions {
  /** Chỉ lấy file có phần mở rộng này (vd `.vue`, `.test.ts`). Bỏ trống = mọi file. */
  ext?: string | readonly string[]
  /**
   * Dân số TỐI THIỂU. Quét ra ít hơn ⇒ NÉM LỖI.
   * Đây là chốt chặn xanh giả — luôn truyền cho mọi lần quét thư mục.
   */
  min: number
  /** Mặc định true (đệ quy xuống thư mục con). */
  recursive?: boolean
  /** Bỏ qua file/thư mục khớp predicate. */
  skip?: (absPath: string) => boolean
}

/**
 * Quét thư mục và **chốt dân số tối thiểu**.
 *
 * Ném lỗi nếu `dir` không tồn tại hoặc số file quét được `< min`, thay vì trả
 * mảng rỗng — mảng rỗng là cách mọi khẳng định "không có vi phạm" trở thành
 * đúng-một-cách-rỗng-tuếch.
 */
export function listFiles(dir: string, opts: ListFilesOptions): string[] {
  const { ext, min, recursive = true, skip } = opts
  if (!existsSync(dir) || !statSync(dir).isDirectory()) {
    throw new Error(
      `[test/paths.ts] listFiles: thư mục không tồn tại: ${dir}\n` +
        'Guard đang trỏ vào hư vô — sửa đường dẫn, KHÔNG hạ ngưỡng min.',
    )
  }
  const exts = ext === undefined ? null : (Array.isArray(ext) ? ext : [ext as string])
  const out: string[] = []

  const walk = (current: string): void => {
    for (const entry of readdirSync(current, { withFileTypes: true })) {
      const abs = resolve(current, entry.name)
      if (skip?.(abs)) continue
      if (entry.isDirectory()) {
        if (recursive) walk(abs)
      } else if (!exts || exts.some((e) => entry.name.endsWith(e))) {
        out.push(abs)
      }
    }
  }
  walk(dir)
  out.sort()

  if (out.length < min) {
    throw new Error(
      `[test/paths.ts] listFiles(${relRepo(dir)}) chỉ ra ${out.length} file, ` +
        `dưới ngưỡng tối thiểu ${min}.\n` +
        'Thư mục đã bị dời/đổi tên/rỗng đi ⇒ guard đã NGỪNG CANH. ' +
        'Sửa đường dẫn hoặc cập nhật ngưỡng CÓ CHỦ Ý — đừng để guard đếm 0 rồi PASS.',
    )
  }
  return out
}
